xbmc/addons/ScreenSaver.cpp

```diff
@@ -53,27 +45,40 @@ CScreenSaver::~CScreenSaver()
 
   delete m_ifc.screensaver->toAddon;
   delete m_ifc.screensaver->toKodi;
-  delete m_ifc.screensaver->props;
   delete m_ifc.screensaver;
 }
 
 bool CScreenSaver::Start()
 {
-  if (m_ifc.screensaver->toAddon->Start)
-    return m_ifc.screensaver->toAddon->Start(m_ifc.screensaver);
+  if (m_ifc.screensaver->toAddon->start)
+    return m_ifc.screensaver->toAddon->start(m_ifc.hdl);
   return false;
 }
 
 void CScreenSaver::Stop()
 {
-  if (m_ifc.screensaver->toAddon->Stop)
-    m_ifc.screensaver->toAddon->Stop(m_ifc.screensaver);
+  if (m_ifc.screensaver->toAddon->stop)
+    m_ifc.screensaver->toAddon->stop(m_ifc.hdl);
 }
 
 void CScreenSaver::Render()
 {
-  if (m_ifc.screensaver->toAddon->Render)
-    m_ifc.screensaver->toAddon->Render(m_ifc.screensaver);
+  if (m_ifc.screensaver->toAddon->render)
+    m_ifc.screensaver->toAddon->render(m_ifc.hdl);
 }
 
-} /* namespace ADDON */
+void CScreenSaver::GetProperties(struct SCREENSAVER_PROPS* props)
+{
+  if (!props)
+    return;
+
+  props->x = 0;
+  props->y = 0;
+  props->device = CServiceBroker::GetWinSystem()->GetHWContext();
+  props->width = CServiceBroker::GetWinSystem()->GetGfxContext().GetWidth();
+  props->height = CServiceBroker::GetWinSystem()->GetGfxContext().GetHeight();
+  props->pixelRatio = CServiceBroker::GetWinSystem()->GetGfxContext().GetResInfo().fPixelRatio;
+  props->name = strdup(Name().c_str());
+  props->presets = strdup(CSpecialProtocol::TranslatePath(Path()).c_str());
```

Comment:
`std::strdup` via &lt;cstring>. Also what I've done in the past, is add a std::string member, and assign its .c_str() to the variable, so that the data is cleaned via RAII

Reply:
Know what you mean, but can not works in all cases.
Reasons:
- In case a addon is in "C", has several instances and stores somewhere his pointer. Then can not be sure that pointer still exists.
  - in a good independent "C" should a `const char*` only be given if it stays in all cases present.
- Another part is about Sandbox where the addon runs independent from Kodi. There can never be a `const char*` given to another target.
  - As the sandbox addon library can not store then itself e.g. a `std::string` to get his "C" characters must be always a string created and then from addon deleted.
  - **Note:** By Sandbox inside function values can be a `const char*` given (as memory then in shared part) only on structures not so easy.

Reply:
Think to remove the three string values.
They are not really needed and can be also done with other callbacks (e.g. `kodi::addon::GetAddonPath()`).